File: packages/tools/ai/src/ai.ts
import fs from "fs";
import path from "path";
import OpenAI from "openai";

const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY || "" });

export async function createTranscription(
  filePath: string,
  resultPath: string,
  fileName: string
): Promise<void> {
  if (!fs.existsSync(filePath)) {
    console.log("File does not exist");
    return;
  }

  if (!fs.existsSync(resultPath)) {
    fs.mkdirSync(resultPath, { recursive: true });
  }

  console.log(`Transcribing ${filePath}...`);
  try {
    const transcription = await openai.audio.transcriptions.create({
      file: fs.createReadStream(filePath),
      model: "whisper-1",
      response_format: "text",
      temperature: 0.3,
    });

    if (transcription) {
      fs.writeFileSync(path.join(resultPath, fileName), String(transcription));
      console.log(`Transcription saved to ${resultPath}`);
    } else {
      console.log("No transcription data received.");
    }
  } catch (error) {
    console.error("Error during transcription:", error);
  }
}

export async function createSummary(
  transcriptedFilePath: string,
  textFilePath: string,
  fileName: string
): Promise<void> {
  if (!fs.existsSync(textFilePath)) {
    fs.mkdirSync(textFilePath, { recursive: true });
  }

  if (!fs.existsSync(transcriptedFilePath)) {
    console.log("File does not exist");
    return;
  }

  const input_text = fs.readFileSync(transcriptedFilePath);

  const response = await openai.chat.completions.create({
    model: "gpt-3.5-turbo-1106",
    messages: [
      {
        role: "system",
        content:
          "You, as an AI, have experience in the crypto field. Try to avoid names. Summarize this text given by the user. It will be delimited by triple single quotes for a description for a video:\\n\\n'''${input_text}'''`.\n\nThe summary is meant for users to understand what the video is about, before watching the actual video.",
      },
      {
        role: "user",
        content: `'''${input_text}'''`,
      },
    ],
    temperature: 0.5,
    max_tokens: 256,
    top_p: 1,
    frequency_penalty: 0.4,
    presence_penalty: 0,
  });

  console.log(response.choices[0].message.content);
  if (!response.choices[0].message.content) {
    console.log("Something went wrong making the summary");
  }

  fs.writeFileSync(
    path.join(textFilePath, fileName),
    String(response.choices[0].message.content)
  );
  console.log(`Transcription saved to ${textFilePath}`);
}

File: packages/tools/ai/src/ai.py
import os
import subprocess
from typing import Any

import whisper
import openai

GPT_MODEL = "gpt-3.5-turbo-1106"
EMBEDDING_MODEL = 'all-MiniLM-L6-v2'
OPENAI_API_KEY = os.getenv('CHATGPT_API_KEY')
if not OPENAI_API_KEY:
    raise EnvironmentError("CHATGPT_API_KEY environment variable not set")


def generate_summary(input_text: str, temperature: float) -> Any:
    response = openai.ChatCompletion.create(
        model=GPT_MODEL,
        messages=[{
            "role":
            "user",
            "content":
            f"You have experience in the crypto field. Summarize this text delimited by triple single quotes for a description for a "
            f"video:\n\n'''{input_text}'''"
        }],
        temperature=temperature,
        max_tokens=512,
        frequency_penalty=0.3,
        presence_penalty=0)
    return response['choices'][0]['message']['content']


def summarize_text(input_text: str, output_file: str) -> None:
    openai.api_key = OPENAI_API_KEY

    # Generate summaries with different temperatures
    summaries = [
        generate_summary(input_text, temp) for temp in [0.3, 0.5, 0.7]
    ]

    summary_input_text = "\n\n\n'''".join(summaries)
    summary_of_summaries = generate_summary(summary_input_text, 0.4)

    with open(output_file, 'w') as file:
        file.write(summary_of_summaries)


def download_m3u8_video_as_wav(m3u8_url: str, output_name: str) -> None:
    command = [
        "ffmpeg", "-analyzeduration", "100M", "-probesize", "100M", "-i",
        m3u8_url, "-vn", "-c:a", "pcm_s16le", "-ar", "44100",
        f'{output_name}.wav'
    ]

    try:
        subprocess.run(command, check=True)
        print(
            f"Successfully downloaded and converted audio from {m3u8_url} to {output_name}.wav"
        )
    except subprocess.CalledProcessError as e:
        print(
            f"Error occurred while downloading and converting audio from {m3u8_url} to {output_name}.wav: {e}"
        )


def convert_speech_to_text(input_wav: str) -> Any:
    model = whisper.load_model("small.en")
    result = model.transcribe(input_wav)
    text = result["text"]

    print("Successfully saved speech-to-text result")
    return text


def start_summary(playbackUrl: str, temp_file_path: str) -> None:
    download_m3u8_video_as_wav(playbackUrl, temp_file_path)
    transcribed_text = convert_speech_to_text(f"{temp_file_path}.wav")
    summarize_text(transcribed_text, f'{temp_file_path}.txt')
